✨ [feat] PoemWritingView, Model 로직 구현 및 UI 구현

Mark3891 opened the pull request:
### Issue Number

close #22


### 🖥️ 작업 내역

> 구현 내용 및 작업 했던 내역을 적어주세요.

1. poemWritingView를 구현했습니다.

- viewModel
```Swift
init(poem: Poem?, coordinator: Coordinator) {
        self.coordinator = coordinator

        if let existingPoem = poem {
            let titleNotEmpty = !existingPoem.title.trimmingCharacters(
                in: .whitespacesAndNewlines
            ).isEmpty
            let contentNotEmpty = !existingPoem.content.trimmingCharacters(
                in: .whitespacesAndNewlines
            ).isEmpty

            self.state = State(
                poem: existingPoem,
                isNew: !(titleNotEmpty || contentNotEmpty)
            )
        } else {
            let newPoem = Poem(
                type: PoemType.none,
                title: "",
                content: "",
                date: Date()
            )
            self.state = State(poem: newPoem, isNew: true)
        }
        updateCanSaveTemporarily()  // 초기값 설정
    }

```
새로 시 작성인지 편집인지를 시 title이나 content 둘 다 없어야 isNew = true로 하여 State를 설정하였습니다.


```Swift
 enum Action {
        case savePoem(context: ModelContext, isCompleted: Bool)
        case updatePoem(context: ModelContext, isCompleted: Bool)
        case updateTitle(String)
        case updateContent(String)
        case updateType(PoemType)
        case updateAlignment(TextAlignmentType)
        case onSheetAppeared(context: ModelContext)
        case onAppeared(context: ModelContext)
        case updateCanSaveTemporarily

    }
```
위의 내용은 viewModel action입니다.


```Swift
case .onAppeared(let context):
            loadSelectedInspirationIfNeeded(context: context)
```
view가 appear할때는 선택된 시상이 있다면 가져오게 하였습니다.


```Swift
 case .onSheetAppeared(let context):
            loadAllInspirations(context: context)
            // Debug
//            state.inspirationList = setTestInspirations()
```
시상 불러오기 sheet가 appear할때는 전체 시상 불러오기를 하였습니다.

```Swift
case .savePoem(let context, let isCompleted):
            state.poem.isCompleted = isCompleted
            let result = SwiftDataManager.shared.savePoem(
                poem: state.poem,
                context: context
            )
            switch result {
            case .success:
                print("시 저장 성공 (\(isCompleted ? "완전 저장" : "임시 저장"))")

                if isCompleted {
                    ///Todo: 나의 시집 (완성)으로 가기
                    coordinator.removeAll()
                } else {
                    coordinator.popLast()
                }
            case .failure(let error):
                print("시 저장 실패:", error.localizedDescription)
            }
        case .updatePoem(let context, let isCompleted):
            state.poem.isCompleted = isCompleted
            let result = SwiftDataManager.shared.updatePoem(context: context)
            switch result {
            case .success:
                print("시 수정 성공 (\(isCompleted ? "완전 저장" : "임시 저장"))")

                if isCompleted {
                    ///Todo: 나의 시집
                    coordinator.removeAll()
                } else {
                    coordinator.popLast()
                }
            case .failure(let error):
                print("시 저장 실패:", error.localizedDescription)
            }
```
임시저장 과 시끝맺기를 isCompleted로 나누어 구현했습니다.
또한 isNew인가에 따라 savePoem함수와 updatePoem함수를 나누었습니다.

```Swift
 @Published var canSaveTemporarily: Bool = false

private func updateCanSaveTemporarily() {
        let titleNotEmpty = !state.poem.title.trimmingCharacters(
            in: .whitespacesAndNewlines
        ).isEmpty
        let contentNotEmpty = !state.poem.content.trimmingCharacters(
            in: .whitespacesAndNewlines
        ).isEmpty
        canSaveTemporarily = titleNotEmpty || contentNotEmpty
    }
```
임시저장 활성화 비활성화를 위해 아래의 canSaveTemporarily를 구현했습니다.
이 값에 따라 임시저장 버튼과 시끝맺기 버튼의 활성화 비활성화를 구분합니다.
로직은 title이나 content 둘 중에 하나라도 값이 있다면 true가 되도록 했습니다.

![Simulator Screen Recording - iPhone 16 Pro - 2025-06-05 at 12 17 47](https://github.com/user-attachments/assets/93f51340-9b25-49a3-96db-3e52990ce79b)

위의 영상은 구현된 UI와 시쓰기 입니다.
지금은 시상 불러오기에서 저장된 시상이 context가 아니라서 다시 눌렀을 때 id가 바뀌기 때문에 checkbox가 반영 안되지만 Swift Data로 하면 됩니다.

Mark3891 committed:
✨ [feat] #22 시상선택버튼 및 시상보기 버튼 구현

- 시상선택버튼 및 시상보기 버튼 구현했습니다.

Mark3891 committed:
✨ [feat] #22 시상선택 시 시상불러오기 시트 구현

- 시상선택 버튼 누를 때 시상 불러오기 시트 구현했습니다.

Mark3891 committed:
✨ [feat] #22 시상보기 버튼 누를 때, 일상 / 감상 읽기 시트 구현

- 시상버튼 누를 시, 일상 / 감상 읽기 시트 페이지 구현했습니다.

Mark3891 committed:
✨ [feat] #22 시 제목 컴포넌트 구현

시제목 컴포넌트 구현했습니다.

Mark3891 committed:
✨ [feat] #22 시 내용 텍스트에디터 구현 및 텍스트 정렬 선택 버튼 구현

- 시 내용 적는 textEditor 구현
- 텍스트 정렬 구현

Mark3891 committed:
✨ [feat] #22 시쓰기 뷰 구현 및 시쓰기 뷰모델 구현

- 네비게이션 바 구현했습니다.
- "시 끝맺기" 버튼 구현했습니다.
- 뷰 모델 임시저장 및 완전 저장 및 업데이트 구현했습니다.

Mark3891 committed:
✨ [feat] #22 시쓰기 버튼 누를 시, 시쓰기 뷰로 네이게이팅 구현

- 시쓰기 버튼 누를 시, 시쓰기 뷰로 네이게이팅 구현했습니다.

Mark3891 committed:
🐛[fix] #22 isNew 판별 기능 구현

- writing view로 poem을 넘겨줘도 title이나 content 둘 중에 하나라도 데이터가 있으면 isNew = false, title이나 content 둘 중에 하나라도 데이터 있으면 isNew = true
- 일상/감상읽기 뷰에서 새로 시쓰기 눌렀을 때 일상/감상 type(id)를 가진 poem이 넘어와야 하기에 위와 같은 로직 구현

Mark3891 committed:
🐛[fix] #22 시 끝맺기 버튼 저장 활성화 로직 수정

- 임시보관 버튼 처럼 제목이나 내용 중 하나라도 입력되어야 저장 버튼 활성화

Mark3891 committed:
✅[Chore] #22 디버기용 삭제

- 데이터 저장 잘 됐는지 확인하기 위해 사용했던 것 삭제
- setInspiration 함수 삭제 및 SwiftDate context 활용한 loadAllInspiration함수로 구현

Mark3891 committed:
Merge branch 'develop' into feat/#22/poem-write-view

Mark3891 committed:
✅ [Chore] #22 DailyReadingSheet 구현 반영

DailyReadingContentView 구현을 반영하여 적용하였습니다.

Mark3891 committed:
✅ [Chore] #22 사용할 필요없는 코딩들 지우기

- popLast()는 이미 save나 update 기능에서 구현되어 있어서 없어도 된다.
- title이 값이 없어도 '무제'를 붙일 필요가 없어서 지웠습니다.

keon22han commented on FunForYou/FunForYou/Sources/Presentation/Poem/PoemWriting/PoemWritingView.swift:
하나의 뷰 struct 내에는 하나의 뷰 레이아웃 컨테이너 (VStack, HStack, ZStack) 을 사용하도록 서브뷰로 나눠주세요!

keon22han replied:
(Feat/#22/PoemWritingView, Model 로직 구현 및 UI 구현 -> ✨ [feat] PoemWritingView, Model 로직 구현 및 UI 구현) 로PR 타이틀 양식 맞춰주시고, Issue Number 에는 Issue 에 대한 작업이 완료되었다면 close 를 앞에 붙여주세요~! 해당 PR은 제가 수정해놓겠습니다.

keon22han commented on FunForYou/FunForYou/Sources/Presentation/Poem/PoemWriting/subView/NoInspirationOptionRow.swift:
여기도 하나의 뷰 struct 내에는 하나의 뷰 레이아웃 컨테이너 (VStack, HStack, ZStack) 을 사용하도록 서브뷰로 나눠주세요!

keon22han commented on FunForYou/FunForYou/Sources/Presentation/Poem/PoemWriting/subView/PoemTitleField.swift:
해당 뷰는 TextField 하나만 들어가있어서 굳이 struct 로 나눌 필요는 없어보입니다! 다음에 참고해주세요!

keon22han commented on FunForYou/FunForYou/Sources/Presentation/Poem/PoemWriting/PoemWritingView.swift:
```diff
@@ -5,18 +5,153 @@
+        ZStack {
+            VStack(spacing: 0) {
+                ScrollView {
+                    VStack(spacing: 0) {
```
VStack 은 바깥에 있어서 안에서는 제거해주는 게 좋아보입니다!

Mark3891 replied:
수정해볼게요~!

Mark3891 committed:
♻️ [refactor] viewModel 중복 부분 삭제

- savePoem과 updatePoem의 중복부분을 없애고 하나의 함수로 구현하여 가독성을 높임
- canSave bool값을 정의하는 부분의 중복부분을 줄이고 하나의 함수로 쓰이기 좋게 함

Mark3891 committed:
🔥 [remove] 디버깅용 print부분 삭제

- 디버깅용 print 부분을 삭제하였습니다.

Mark3891 committed:
♻️ [refactor] : InspirationReadingSheet 추가 및 daily,appreciation sheet 삭제

- 중복되는 파일인 dailyreadingsheet와 appreciationreadingsheet를 하나로 합쳐 appreciationreadingsheet를 만들었습니다.

Mark3891 committed:
♻️ [refactor] : InsprationRowView, NoInspirationCard 추가 및 관련 파일 삭제

- NoInspirationRowView를 삭제하고 비슷한 형태인 InspirationRowView에 inpiration이 없는 경우를 포함하여 하나의 파일로 구성했습니다.
- InspirationRow를 반영하기 위해 InspirationSelectionSheet 파일을 수정했습니다.

Mark3891 committed:
🔥 [remove] 불필요한 파일 삭제

- 가독성을 위해 정리한 파일들을 삭제하였습니다.
- AppriciationReadingSheet, DailyReadingSheet, NoInsprationOptionRow

Mark3891 commented:
중복되는 코드 및 파일 정리하였습니다

keon22han commented on FunForYou/FunForYou/Sources/Presentation/Poem/PoemWriting/subView/TextAlignmentPicker.swift:
```diff
@@ -0,0 +1,63 @@
+                .foregroundColor(FFYColor.gray3)
+                .padding(.trailing, 20)
+
+            HStack {
```
바깥 HStack 뷰 레이아웃 컨테이너가 있어서, 안쪽에는 제거해도 될 것 같아요!

Mark3891 committed:
✅ [chore] 설명 추가 및 버그 수정

- InspirationSelectionSheet에서 설명 필요부분 적어주었습니다.
- InspirationReadingSheet에서 제목부분 오류 부분 수정해주었습니다.

Mark3891 committed:
✅ [chore] 하나의 뷰 struct 내에는 하나의 뷰 레이아웃 컨테이너 (VStack, HStack, ZStack) 을 사용하도록 서브뷰로 나눔

PoemWritingView
- 쓸데없는 VStack을 제거했습니다.
- ScrollView 코드 부분을 component화해서 파일 생성했습니다.
- Text title도 isEditorFocused를 적용하여 외부 부분 탭 시 키보드가 내려가도록 설정해두었습니다.

Mark3891 committed:
✅ [chore] TextField component화 제거

- PoemTitleField 뷰 제거
- PoemWritingContentView, textField로 수정

Mark3891 committed:
✅ [chore] TextAlignmnetPicker 불필요한 HStack 제거

- TextAlignmnetPicker 불필요한 HStack 제거했습니다.

Mark3891 commented:
@keon22han 반영해서 다 수정했습니다~!
